BlueprintUI/Sources/Internal/ElementIdentifier.swift

```diff
@@ -1,43 +1,114 @@
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 1)
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 2)
+ ElementC -> ElementIdentifier(ElementC.self, [optional key], 1)
+ ```
+ 
+ And then removed A and the first B:
+ 
+ ```
+ ElementB -> ElementIdentifier(ElementB.self, [optional key], 1)
+ ElementC -> ElementIdentifier(ElementC.self, [optional key], 1)
+ ```
+ 
+ You will note that the identifiers remain stable, which ultimately ensures that views are reused.
+ */
+struct ElementIdentifier: Hashable, CustomDebugStringConvertible {
+    
+    let elementType : ObjectIdentifier
+    let key : AnyHashable?
 
-    internal init(key: String?, index: Int) {
-        if let key = key {
-            self = .key(key)
+    let count : Int
+
+    init(elementType : Element.Type, key : AnyHashable?, count : Int) {
+        
+        self.elementType = ObjectIdentifier(elementType)
+        self.key = key
+        
+        self.count = count
+    }
+    
+    var debugDescription: String {
+        if let key = self.key {
+            return "\(self.elementType).\(String(describing: key)).\(self.count)"
         } else {
-            self = .index(index)
+            return "\(self.elementType).\(self.count)"
         }
     }
-
-    /// Returns the reuse identifier of this component (if provided).
-    var key: String? {
-        switch self {
-        case .key(let key):
-            return key
-        default:
-            return nil
+    
+    /**
+     Internal type used to create `ElementIdentifier` instances during view hierarchy updates.
+     */
+    struct Factory {
+        
+        init(reserveCapacity capacity: Int? = nil) {
+            
+            self.countsByKey = [:]
+            
+            if let capacity = capacity {
+                self.countsByKey.reserveCapacity(capacity)
+            }
         }
-    }
-
-}
-
-extension ElementIdentifier: CustomStringConvertible {
-
-    var description: String {
-        switch self {
-        case let .key(string):
-            return string
-        case let .index(index):
-            return "(\(index))"
+                
+        mutating func nextIdentifier(for type : Element.Type, key : AnyHashable?) -> ElementIdentifier {
+            
+            let count = self.nextCount(for: type, key: key)
+            
+            return ElementIdentifier(
+                elementType: type,
+                key: key,
+                count: count
+            )
+        }
+        
+        private var countsByKey : [Key:Int]
+        
+        mutating private func nextCount(for type : Element.Type, key : AnyHashable?) -> Int {
+            
+            let key = Key(
+                elementType: ObjectIdentifier(type),
+                key: key
+            )
+            
+            let current = self.countsByKey[key, default: 1]
+            
+            self.countsByKey[key] = (current + 1)
```

Comment:
If I understand correctly, key counts only go up right? If an element is removed, the count doesn't decrease? If that's true, just wondering if we should worry about overflow.

Reply:
This is per element tree diff, so unless a element has 2^32 children (or whatever), this won't be a problem in practice.